aspnetcore/client-side/import-export-interop.md

```diff
@@ -0,0 +1,271 @@
+```
+
+In the imported method signature, you can use .NET types for parameters and return values, which are marshalled automatically by the runtime. Use `JSMarshalAsAttribute<T>` to control how the imported method parameters are marshalled. For example, you might choose to marshal a `long` as <xref:System.Runtime.InteropServices.JavaScript.JSType.Number?displayProperty=nameWithType> or <xref:System.Runtime.InteropServices.JavaScript.JSType.BigInt?displayProperty=nameWithType>. You can pass <xref:System.Action>/<xref:System.Func%601> callbacks as parameters, which are marshalled as callable JS functions. You can pass both JS and managed object references, and they are marshaled as proxy objects, keeping the object alive across the boundary until the proxy is garbage collected. You can also import and export asynchronous methods with a <xref:System.Threading.Tasks.Task> result, which are marshaled as [JS promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). Most of the marshalled types work in both directions, as parameters and as return values, on both imported and exported methods.
+
+To export a .NET method so it can be called from JS, use the `JSExportAttribute`.
+
+In the following example, the `Greeting` method returns a string that includes the result of calling the `GetHRef` method. As shown earlier, the `GetHref` C# method calls into JS for the `window.location.href` function from the `main.js` module. `window.location.href` returns the current page address (URL). In an app generated from either the `wasmbrowser` or `wasmconsole` templates, `Greeting` is in the `Program.cs` file with the partial class namespace `MyClass`:
+
+```csharp
+[JSExport]
+internal static string Greeting()
+{
+    var text = $"Hello, World! Greetings from {GetHRef()}";
+    Console.WriteLine(text);
+    return text;
+}
+```
+
+In the `index.html` file of an app generated from the `wasmbrowser` template, the `main.js` and `dotnet.js` modules are preloaded. The `<script>` tag for `main.js` appears inside the closing `</body>` tag, resulting the following:
+
+* `main.js` is loaded.
+* `dotnet.js` is imported by `main.js`.
+* The .NET runtime is started by calling `runMainAndExit()` in `main.js`.
+
+<!--
+
+    NOTE: In the preceding list, change runMainAndExit() to dotnet.run() for release.
+
+-->
+
+```html
+<html>
+
+<head>
+    ...
+    <link rel="modulepreload" href="./main.js" />
+    <link rel="modulepreload" href="./dotnet.js" />
+</head>
+
+<body>
+    ...
+    <script type='module' src="./main.js"></script>
+</body>
+
+</html>
+```
+
+In addition to preloading, it's possible to prefetch the largest binary files. In a production app, measure the impact of preloading and prefetching optimizations. For more information, see [Use .NET from any JavaScript app in .NET 7: Optimize the app](https://devblogs.microsoft.com/dotnet/use-net-7-from-any-javascript-app-in-net-7/#optimize-the-app).
```

Comment:
This section is pretty specific to how the templates are setup. For most users I expect they will have their own host HTML file and they can decide how they want to optimize how the app loads. I don't think we really need to go through the specific HTML file from the project template.

However, we do need to document where to find *dotnet.js* so that users can import it. This file will be generated as part of the build output of the app in the AppBundle folder. To integrate with an existing app, the user should copy the contents of the AppBundle folder so that it can be served along with the rest of the app. For production deployments, they should do a `dotnet publish -c Release` and deploy the AppBundle folder contents from the publish output.

```suggestion
```